File: 408_add-binary/add-binary.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/add-binary
@Language: Python
@Datetime: 15-05-19 04:52
'''

class Solution:
    # @param {string} a a number
    # @param {string} b a number
    # @return {string} the result
    def addBinary(self, a, b):
        # Write your code here
        res, i, j, carry = [], len(a) - 1, len(b) - 1, 0
        while i >= 0 or j >= 0 or carry > 0:
            v1, v2 = 0, 0
            if i >= 0: v1 = int(a[i])
            if j >= 0: v2 = int(b[j])
            sum = v1 + v2 + carry
            res.append(str(sum % 2))
            i, j, carry = i - 1, j - 1, sum / 2 
        
        return ''.join(reversed(res))
        
        

File: 408_add-binary/add-binary.cpp
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/add-binary
@Language: C++
@Datetime: 15-05-19 04:42
*/

class Solution {
public:
    /**
     * @param a a number
     * @param b a number
     * @return the result
     */
    string addBinary(string& a, string& b) {
        // Write your code here
        string res;
        int i = a.size() - 1, j = b.size() - 1, carry = 0;
        for (; i >= 0 && j >= 0; i--, j--) {
            int sum = a[i] - '0' + b[j] - '0' + carry;
            res.push_back(sum % 2 + '0');
            carry = sum / 2;
        }
        
        for (; i >= 0; i--) {
            int sum = a[i] - '0'  + carry;
            res.push_back(sum % 2 + '0');
            carry = sum / 2;
        }
        
        for (; j >= 0; j--) {
            int sum = b[j] - '0'  + carry;
            res.push_back(sum % 2 + '0');
            carry = sum / 2;
        }
        
        if (carry > 0)
            res.push_back(carry + '0');
            
        return string(res.rbegin(), res.rend());
    }
};
